client/my-sites/controller.js

```diff
@@ -381,9 +382,22 @@ export function siteSelection( context, next ) {
 			return;
 		}
 	} else {
+		// Fetch the site from siteFragment.
+		dispatch( requestSite( siteFragment ) )
+			.then( () => {
```

Comment:
The promise callback here is almost identical to function `selectOnSitesChange` and serves the same purpose: finish the selection asynchronously after the site is fetched. Do you think you could use it here?

Reply:
Yeah, I know. However, I worry that if the first condition in `selectOnSitesChange` doesn't pass, it will show the "Finishing set up..." bug and request all sites again.

Reply:
Do you have any idea why the "Finishing set up..." code is there at all? It's currently broken, because the two counts don't match if you have any inactive sites. So it should be either fixed, or removed.

What it seems to do is that if the site cannot be found even after trying to fetch it from the server, it requests the site list again and tries once more before giving up. While this second request is being executed, it displays the "Finishing set up..." notice.

It looks like it's a part of signup -- the site may not be created yet at the time when we try to load a `/something/new.site.slug` route.

I don't want to distract you from your main task, but it would be nice to spend some limited time with some history research and find who added the code and why -- or ask someone who knows, like @mtias :)

Reply:
I looked back through the history and saw that it was added in #7250. It was fixing the following bug: #5571.

If I understand it correctly, people used to connect to WP.com through Jetpack UI in `wp-admin` back then. So, you could connect the site but not having your local Calypso cache refreshed. Then, if you visited some direct link from wp-admin's Jetpack UI into Calypso, you would see all sites view as the newly connected JP site was not there.

However, I just tested and when I connected my `.wpsandbox` site now from wp-admin, it took me through Calypso, not through wp-admin's Jetpack menu. Thus, I think it's safe to remove the code and reuse the function!

/cc @ryelle @johnHackworth to confirm.

Reply:
Hi there!

I'm afraid it's not safe. You could be using your jetpack site from a different computer and click on any of the links in wp-admin that takes you to a (clean-slate in that computer) calypso. You could also have connected your site before we used calypso for it and use one of those links. 

I'm sure there could be better solutions, but as far as I know, removing this, ATM, would break wp.com experience for some Jetpack users :/

Reply:
Hey @johnHackworth, thanks for chiming in!

> You could also have connected your site before we used calypso for it and use one of those links.

That's not a problem, though.

Normally, we always fetch all the sites with `/me/sites` endpoint in the [layout](https://github.com/Automattic/wp-calypso/blob/522b2790dcdf2e9fcdd367ff68e2c3fc8c76d540/client/layout/index.jsx#L165) component when you load wp.com's any page. So, if you connected a site long time ago, not even in Calypso, and you load Calypso now for the first time, the site will get fetched with the initial `/me/sites` request.

According to #7250, the problem it was solving was that: you connected through wp-admin and then immediately clicked on some link which took you to calypso. @ryelle wrote that
> the me/sites endpoint can have a delay before registering that the user has a new site
so in this case, you connected the site outside of calypso and requested `/me/sites` too soon after connecting the site when navigated to calypso.

So, this definitely doesn't apply to your second point since I assume the delay until `/me/sites` registers that you have a new site is not days, not even minutes, maybe seconds.

But what about:
> You could be using your jetpack site from a different computer and click on any of the links in wp-admin that takes you to a (clean-slate in that computer) calypso.

That's also a non issue, see my comment above. However, your thought brought me the following scenario:

What if the customer used one computer to connect their JP site and then, after connecting it, immediately tried to open Calypso on another computer? Well, depending on how long it takes `/me/sites` to register that a site was connected, if the customer is fast enough, they could replicate the bug.

Buuut, what are we talking about here? 😆  Do you know somebody who will have two computers ready, click "connect" on one and then immediately reload calypso on the other one so they can complain?

Even if it takes a minute for `/me/sites` to register that you have connected a new site, why would you switch devices so fast just right after connecting a brand new JP site? Wouldn't you explore it first, and then maybe in 10 minutes, when on bus, use your phone to access (which should be fine).

I'll try to find out how long it takes for `/me/sites` to return the newly connected site. If it's under a minute, I'm fine with that.

Reply:
Just tested and `/me/sites` returns the newly connected JP site in under 5 seconds. I think we are fine here. :) @johnHackworth if you can get us to a breaking situation, let me know please. I'd like to merge this tomorrow.

Reply:
oh, ok... I guess all this changed quite a while since the last time I really experimented with it... so... ok, I have no other worries about this PR! :)